src/sentry/sentry_metrics/multiprocess.py

```diff
@@ -0,0 +1,334 @@
+from arroyo.processing.strategies.streaming.transform import ParallelTransformStep
+from arroyo.types import Message, Partition, Position, Topic
+from django.conf import settings
+
+from sentry.utils import json, kafka_config
+
+TPayload = TypeVar("TPayload")
+TProcessed = TypeVar("TProcessed")
+
+DEFAULT_QUEUED_MAX_MESSAGE_KBYTES = 50000
+DEFAULT_QUEUED_MIN_MESSAGES = 10000
+
+logger = logging.getLogger(__name__)
+
+
+def initializer():
+    from sentry.runner import configure
+
+    configure()
+
+
+def get_config(topic: str, **options) -> MutableMapping[str, Any]:
+    consumer_config = kafka_config.get_kafka_consumer_cluster_options(
+        "default",
+        override_params={
+            "enable.auto.commit": False,
+            "enable.auto.offset.store": False,
+            "group.id": "ingest-metrics-consumer",
+            "default.topic.config": {"auto.offset.reset": "latest"},
+            # overridden to reduce memory usage when there's a large backlog
+            "queued.max.messages.kbytes": DEFAULT_QUEUED_MAX_MESSAGE_KBYTES,
+            "queued.min.messages": DEFAULT_QUEUED_MIN_MESSAGES,
+        },
+    )
+    return consumer_config
+
+
+class MetricsBatchBuilder(Generic[TPayload]):
+    def __init__(self, max_batch_size: int, max_batch_time: float) -> None:
+        self.__messages = []
+        self.__max_batch_size = max_batch_size
+        self.__deadline = time.time() + max_batch_time
+
+    def __len__(self) -> int:
+        return len(self.__messages)
+
+    @property
+    def messages(self):
+        return self.__messages
+
+    def append(self, message: Message[TPayload]) -> None:
+        self.__messages.append(message)
+
+    def ready(self) -> bool:
+        if len(self.messages) >= self.__max_batch_size:
+            return True
+        elif time.time() >= self.__deadline:
+            return True
+        else:
+            return False
+
+
+class BatchMessages(ProcessingStep):
+    def __init__(
+        self,
+        next_step: ProcessingStrategy[TProcessed],
+        max_batch_time: float,
+        max_batch_size: int,
+    ):
+        self.__max_batch_size = 2
+        self.__max_batch_time = max_batch_time
+
+        self.__next_step = next_step
+        self.__batch: Optional[MetricsBatchBuilder] = None
+        self.__closed = False
+
+    def poll(self) -> None:
+        assert not self.__closed
+
+        self.__next_step.poll()
+
+        if self.__batch and self.__batch.ready():
+            self.__flush()
+
+    def submit(self, message: Any) -> None:
+        if self.__batch is None:
+            self.__batch = MetricsBatchBuilder(self.__max_batch_size, self.__max_batch_time)
+
+        self.__batch.append(message)
+
+    def __flush(self) -> None:
+        if not self.__batch:
+            return
+        last = self.__batch.messages[-1]
+
+        new_message = Message(last.partition, last.offset, self.__batch.messages, last.timestamp)
+        self.__next_step.submit(new_message)
```

Comment:
This is not going to typecheck. 
You declared the next step as 
```
next_step: ProcessingStrategy[TProcessed],
```
but you are passing it a `Message`.
The type checker cannot know if TProcessed is actually the same type as Message. According to the type checker you are trusting the caller of the constructor to choose the type of `next_step`, though the only thing it is going to enforce is that if `ProcessingStrategy[TProcessed],` is passed to the constructor, the submit call must receive `TProcessed`. So you cannot just pass message here. You need to pass `TProcessed` which you do not have a way to build because the `messages` method in the batch is untyped. You need to give it a type that is compatible with `TProcessed`.

Welcome to generics :D (happy to go through this in person if you want).